File: app/src/main/java/com/app/missednotificationsreminder/settings/SettingsViewModel.java
package com.app.missednotificationsreminder.settings;

import android.Manifest;
import android.content.Context;
import android.os.Vibrator;
import android.text.TextUtils;

import com.app.missednotificationsreminder.binding.model.BaseViewModel;
import com.app.missednotificationsreminder.binding.util.BindableBoolean;
import com.app.missednotificationsreminder.binding.util.BindableString;
import com.app.missednotificationsreminder.di.qualifiers.FragmentScope;
import com.app.missednotificationsreminder.service.ReminderNotificationListenerService;
import com.app.missednotificationsreminder.service.util.ReminderNotificationListenerServiceUtils;
import com.app.missednotificationsreminder.util.BatteryUtils;
import com.tbruyelle.rxpermissions.RxPermissions;

import javax.inject.Inject;

import rx.Observable;
import timber.log.Timber;

/**
 * The view model for the settings view
 *
 * @author Eugene Popovich
 */
@FragmentScope
public class SettingsViewModel extends BaseViewModel {

    /**
     * Permissions required by the application
     */
    static String[] REQUIRED_PERMISSIONS = new String[]{
            Manifest.permission.WAKE_LOCK,
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.VIBRATE,
    };

    /**
     * Data binding field used for the access initialized flag
     */
    public BindableBoolean accessInitialized = new BindableBoolean(false);
    /**
     * Data binding field used for the access enabled flag
     */
    public BindableBoolean accessEnabled = new BindableBoolean(false);
    /**
     * Data binding field used for the battery optimization disabled information flag
     */
    public BindableBoolean batteryOptimizationDisabled = new BindableBoolean(false);
    /**
     * Data binding field used for the show advanced settings flag
     */
    public BindableBoolean advancedSettingsVisible = new BindableBoolean(false);
    /**
     * Data binding field used to hold information about whether the vibration is available
     */
    public BindableBoolean vibrationSettingsVisible = new BindableBoolean(false);
    /**
     * Data binding field used to hold information about missing required permissions
     */
    public BindableString missingPermissions = new BindableString();
    /**
     * The vibrator instance
     */
    final Vibrator mVibrator;

    @Inject public SettingsViewModel(Vibrator vibrator) {
        Timber.d("SettingsViewModel: init");
        mVibrator = vibrator;
    }

    /**
     * Run the operation to check whether the notification service is enabled
     */
    public void checkServiceEnabled(Context context) {
        monitor(
                Observable
                        .just(ReminderNotificationListenerService.class)
                        .map(serviceClass -> ReminderNotificationListenerServiceUtils.isServiceEnabled(context, serviceClass))
                        .subscribe(enabled -> {
                            accessEnabled.set(enabled);
                            accessInitialized.set(true);
                        }, t -> Timber.e(t, "Unexpected"))
        );
    }

    /**
     * Run the operation to check whether the battery optimization is disabled for the application
     */
    public void checkBatteryOptimizationDisabled(Context context) {
        batteryOptimizationDisabled.set(!isBatteryOptimizationSettingsVisible() ||
                BatteryUtils.isBatteryOptimizationDisabled(context));
    }

    /**
     * Check whether all required permissions are granted
     */
    public void checkPermissions(Context context) {
        monitor(Observable
                .from(REQUIRED_PERMISSIONS)
                .filter(permission -> !RxPermissions.getInstance(context).isGranted(permission))
                .toList()
                .map(permissions -> TextUtils.join(", ", permissions))
                .subscribe(missingPermissions.asAction()));
    }

    /**
     * Check whether the vibration is allowed on device
     */
    public void checkVibrationAvailable() {
        vibrationSettingsVisible.set(mVibrator.hasVibrator());
    }

    /**
     * Grant required permissions
     */
    public void grantRequiredPermissions(Context context) {
        Timber.d("grantRequiredPermissions");
        monitor(RxPermissions
                .getInstance(context)
                .requestEach(REQUIRED_PERMISSIONS)
                .filter(permission -> !permission.granted) // skip already granted permissions
                .map(permission -> permission.name) // collect names of the not yet granted permissions
                .toList()
                .map(permissions -> TextUtils.join(", ", permissions))
                .subscribe(missingPermissions.asAction()));
    }

    /**
     * Data binding method used to determine whether to display battery optimization settings
     */
    public boolean isBatteryOptimizationSettingsVisible() {
        return BatteryUtils.isBatteryOptimizationSettingsAvailable();
    }
}

File: app/src/main/java/com/app/missednotificationsreminder/settings/SettingsViewModel.kt
package com.app.missednotificationsreminder.settings

import android.Manifest
import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Vibrator
import androidx.core.content.ContextCompat
import androidx.lifecycle.viewModelScope
import com.app.missednotificationsreminder.binding.model.BaseViewStateModel
import com.app.missednotificationsreminder.binding.util.bindWithPreferences
import com.app.missednotificationsreminder.data.model.NightMode
import com.app.missednotificationsreminder.payment.ObservesPendingPayments
import com.app.missednotificationsreminder.payment.ObservesPendingPaymentsImpl
import com.app.missednotificationsreminder.payment.billing.data.source.PurchaseRepository
import com.app.missednotificationsreminder.payment.data.model.Purchase
import com.app.missednotificationsreminder.service.ReminderNotificationListenerService
import com.app.missednotificationsreminder.service.util.ReminderNotificationListenerServiceUtils
import com.app.missednotificationsreminder.settings.di.qualifiers.ForceWakeLock
import com.app.missednotificationsreminder.settings.di.qualifiers.RateAppClicked
import com.app.missednotificationsreminder.util.BatteryUtils
import com.tfcporciuncula.flow.Preference
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject

/**
 * The view model for the settings view
 */
@ExperimentalCoroutinesApi
class SettingsViewModel @Inject constructor(private val vibrator: Vibrator,
                                            private val nightMode: Preference<NightMode>,
                                            @param:ForceWakeLock private val forceWakeLock: Preference<Boolean>,
                                            @param:RateAppClicked private val rateAppClicked: Preference<Boolean>,
                                            private val purchaseRepository: PurchaseRepository,
                                            private val purchases: Preference<List<Purchase>>) :
        BaseViewStateModel<SettingsViewState, SettingsViewStatePartialChanges>(SettingsViewState()),
        ObservesPendingPayments by ObservesPendingPaymentsImpl(purchaseRepository, purchases) {

    init {
        Timber.d("SettingsViewModel: init")
        viewModelScope.launch {
            launch {
                _viewState.bindWithPreferences(nightMode,
                        { newValue, vs ->
                            SettingsViewStatePartialChanges.NightModeChanged(newValue).reduce(vs)
                        },
                        { it.nightMode })
            }
            launch {
                _viewState.bindWithPreferences(forceWakeLock,
                        { newValue, vs ->
                            SettingsViewStatePartialChanges.ForceWakeLockChange(newValue).reduce(vs)
                        },
                        { it.forceWakeLock })
            }
            observePendingPayments(initialDelay = 0)
        }
    }

    /**
     * Run the operation to check whether the notification service is enabled
     */
    fun checkServiceEnabled(context: Context) {
        ReminderNotificationListenerServiceUtils.isServiceEnabled(context, ReminderNotificationListenerService::class.java)
                .run {
                    process(SettingsViewStatePartialChanges.AccessEnabledChange(this))
                }
    }

    /**
     * Run the operation to check whether the battery optimization is disabled for the application
     */
    fun checkBatteryOptimizationDisabled(context: Context) {
        (!viewState.value.isBatteryOptimizationSettingsVisible ||
                BatteryUtils.isBatteryOptimizationDisabled(context))
                .run { process(SettingsViewStatePartialChanges.BatteryOptimizationDisabledChanged(this)) }
    }

    /**
     * Check whether all required permissions are granted
     */
    fun checkPermissions(context: Context) {
        REQUIRED_PERMISSIONS
                .filter {
                    ContextCompat.checkSelfPermission(
                            context,
                            it
                    ) == PackageManager.PERMISSION_DENIED
                }
                .joinToString(", ")
                .run {
                    process(SettingsViewStatePartialChanges.MissingPermissionsChanged(this))
                }
    }

    /**
     * Check whether the vibration is allowed on device
     */
    fun checkVibrationAvailable() {
        vibrator.hasVibrator()
                .run { process(SettingsViewStatePartialChanges.VibrationSettingsAvailableChanged(this)) }
    }

    fun forceWakeLockChanged(value: Boolean) {
        process(SettingsViewStatePartialChanges.ForceWakeLockChange(value))
    }

    fun rateApp(activity: Activity) {
        rateAppClicked.set(true)
        val uri = Uri.parse("market://details?id=${activity.packageName}")
        val goToMarket = Intent(Intent.ACTION_VIEW, uri)
        // To count with Play market backstack, After pressing back button,
        // to taken back to our application, we need to add following flags to intent.
        goToMarket.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY or
                Intent.FLAG_ACTIVITY_MULTIPLE_TASK)
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
            goToMarket.addFlags(Intent.FLAG_ACTIVITY_NEW_DOCUMENT)
        }
        try {
            activity.startActivity(goToMarket)
        } catch (e: ActivityNotFoundException) {
            activity.startActivity(Intent(Intent.ACTION_VIEW,
                    Uri.parse("https://play.google.com/store/apps/details?id=${activity.packageName}")))
        }
    }

    companion object {
        /**
         * Permissions required by the application
         */
        val REQUIRED_PERMISSIONS = listOf(
                Manifest.permission.WAKE_LOCK,
                Manifest.permission.READ_EXTERNAL_STORAGE,
                Manifest.permission.VIBRATE)
    }
}